server/webhook_parser.go

```diff
@@ -0,0 +1,253 @@
+// See License for license information.
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+var webhookWrapperFunc func(wh Webhook) Webhook
+
+func ParseWebhook(in io.Reader) (Webhook, *JiraWebhook, error) {
+	jwh := &JiraWebhook{}
+	err := json.NewDecoder(in).Decode(&jwh)
+	if err != nil {
+		return nil, nil, err
+	}
+	if jwh.WebhookEvent == "" {
+		return nil, jwh, errors.New("No webhook event")
+	}
+	if jwh.Issue.Fields == nil {
+		return nil, jwh, errors.New("Invalid webhook event")
+	}
+
+	wh := func() Webhook {
+		switch jwh.WebhookEvent {
+		case "jira:issue_created":
+			return parseWebhookCreated(jwh)
+		case "jira:issue_deleted":
+			return parseWebhookDeleted(jwh)
+		case "jira:issue_updated":
+			switch jwh.IssueEventTypeName {
+			case "issue_assigned":
+				return parseWebhookAssigned(jwh)
+			case "issue_updated", "issue_generic":
+				return parseWebhookChangeLog(jwh)
+			}
+		case "comment_created":
+			return parseWebhookCommentCreated(jwh)
+		case "comment_updated":
+			return parseWebhookCommentUpdated(jwh)
+		case "comment_deleted":
+			return parseWebhookCommentDeleted(jwh)
+		}
+		return nil
+	}()
+	if wh == nil {
+		return nil, jwh, errors.Errorf("Unsupported webhook data: %v", jwh.WebhookEvent)
+	}
+
+	// For HTTP testing, so we can capture the output of the interface
+	if webhookWrapperFunc != nil {
+		wh = webhookWrapperFunc(wh)
+	}
+
+	return wh, jwh, nil
+}
+
+func parseWebhookChangeLog(jwh *JiraWebhook) Webhook {
+	for _, item := range jwh.ChangeLog.Items {
```

Comment:
We only ever pay attention to the first change-log item we can parse. Is this the indented behavior? We are kinda locking our selves a bit in with this architecture.

Reply:
A separate ticket?

2/5 this is not an architecture issue. a 1-1 mapping of `JiraWebhook -> `Webhook` makes sense. I agree that we should scan all records, and include all recognized changes in the post, or potentially even make multiple posts if applicable. That would be a part of `PostToChannel`.

Reply:
Sure make a ticket.

Reply:
https://mattermost.atlassian.net/browse/MM-15927